packages/opencensus-core/src/metrics/gauges/gauge.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {validateArrayElementsNotNull, validateNotNull} from '../../common/validations';
+import {LabelKey, LabelValue, Metric, MetricDescriptor, MetricDescriptorType, TimeSeries, Timestamp} from '../export/types';
+import * as types from '../gauges/types';
+import {hashLabelValues, initializeDefaultLabels} from '../utils';
+
+/**
+ * Gauge metric
+ */
+export class Gauge implements types.Meter {
+  private readonly metricDescriptor: MetricDescriptor;
+  private labelKeysLength: number;
+  private defaultLabelValues: LabelValue[];
+  private registeredPoints: Map<string, types.Point> = new Map();
+
+  constructor(
+      name: string, description: string, unit: string,
+      type: MetricDescriptorType, readonly labelKeys: LabelKey[]) {
+    this.metricDescriptor = {name, description, unit, type, labelKeys};
+    this.labelKeysLength = labelKeys.length;
+    this.defaultLabelValues = initializeDefaultLabels(this.labelKeysLength);
+  }
+
+  /**
+   * Creates a TimeSeries and returns a Point if the specified
+   * labelValues is not already associated with this gauge, else returns an
+   * existing Point.
+   *
+   * It is recommended to keep a reference to the Point instead of always
+   * calling this method for manual operations.
+   *
+   * @param {LabelValue[]} labelValues The list of the label values.
+   * @returns {types.Point} The value of single gauge.
+   */
+  getOrCreateTimeSeries(labelValues: LabelValue[]): types.Point {
+    validateArrayElementsNotNull(
+        validateNotNull(labelValues, 'labelValues'), 'labelValue');
```

Comment:
Should we show to user some message in case if its `null`? The same question for other `trow new Error` in this PR